### ISSUE TITLE
Profiling: Integrate event logs from application processes besides the IDE

### ISSUE THREAD

**Issue by wdanilo** [enso-org/enso#4300]
This task is automatically imported from the old Task Issue Board and it was originally created by Keziah Wesley.
Original issue is [here](https://www.pivotaltracker.com/story/show/181837344).
<hr />

To troubleshoot backend performance issues effectively, Enso developers need more information about what the backend is doing; specifically, we would like to know when the Language Server is doing work, vs. waiting for the compiler. As an easily-implemented way to shed some light on that distinction, we can log types of messages passed between the two processes, with their timings.

The backend team is working on gathering the data; to display it, we'll need to implement two things in the profiler tools:
- The ability to show metadata from different origins in the graphs, and distinguish their origins visually (this is tracked in a separate issue).
- The ability to attach backend data to a profile file, and make it available to the graphing tools.

## API extension for profile-merging tools to use

```rust
// in profiler::data:

impl<M> Profile<M> {
    fn attach<N>(&mut self, sub: Profile<N>) where N: Into<M> { .. }
    fn attach_with_metadata_mapper<F, N>(&mut self, sub: Profile<N>, f: F) where
        F: FnMut(N) -> M { .. }
}
```

## New type for "core" metadata

```rust
// in profiler

/// Basic metadata that may be present in any profile file.
#[derive(Serialize, Deserialize)]
pub enum CoreMetadata {
    /// Offset that can be added to a `Timestamp` to produce a system time.
    ///
    /// For web `performance.now` timestamps, this is the system time of the *time origin*.
    /// For timestamps that are already in system time, this is 0.
    TimeOffset(/* type TBD */),
}
```

## Tasks:
- [ ] Add to profiling logs metadata establishing relationship between system times and time-origin-based times.
- [ ] Implement a tool that attaches auxiliary logs to a profile file.

## Comments:
**Keziah Wesley** reports a new **STANDUP** for yesterday (2022-04-07):

**Progress:** Worked out API changes and tooling design for merging profiling data from different processes, in order to have better data visualizations ready before the backend-performance meeting on Wednesday. Started implementing design for time-origin information in profiles. It should be finished by 2022-04-12.

**Next Day:** Next day I will be working on the [##181837344 task](https://www.pivotaltracker.com/story/show/181837344). Finish implementing absolute-times; implement the multi-process-profile APIs that will support the merge tool; maybe write the merge tool. (Enso Bot - Apr 8, 2022)
<hr />
**Keziah Wesley** reports a new **STANDUP** for today (2022-04-08):

**Progress:** Worked out API changes and tooling design for merging profiling data from different processes, in order to have better data visualizations ready before the backend-performance meeting on Wednesday. Started implementing design for time-origin information in profiles. It should be finished by 2022-04-12.

**Next Day:** Next day I will be working on the [##181837344 task](https://www.pivotaltracker.com/story/show/181837344). Finish implementing absolute-times; implement the multi-process-profile APIs that will support the merge tool; maybe write the merge tool. (Enso Bot - Apr 8, 2022)
<hr />
**Keziah Wesley** reports a new **STANDUP** for today (2022-04-11):

**Progress:** Almost finished multiprocess profiles. It should be finished by 2022-04-12.

**Next Day:** Next day I will be working on the [##181837344 task](https://www.pivotaltracker.com/story/show/181837344). Finish multiprocess profiles. Get back to batch mode, and hopefully finish it tomorrow too. (Enso Bot - Apr 11, 2022)
<hr />
**Keziah Wesley** reports a new **🔴 DELAY** for yesterday (2022-04-12):

**Summary:** There is **3 days** delay in implementation of the [Profiling: Integrate event logs from application processes besides the IDE (#181837344) task](https://www.pivotaltracker.com/story/show/181837344).
It **will cause 0 days delay** for the delivery of this weekly plan.

**Delay Cause:** Realized profile format backward-compatibility will become important once upcoming batch mode PR is merged and we start sharing profile files, so I'm making some changes to the format that would be painful to do later. (Enso Bot - Apr 14, 2022)
<hr />
**Keziah Wesley** reports a new **STANDUP** for yesterday (2022-04-12):

**Progress:** Multiprocess profile format implemented, backend message log importing tool working. Code needs some cleanup before PR. It should be finished by 2022-04-15.

**Next Day:** Next day I will be working on the [##181837344 task](https://www.pivotaltracker.com/story/show/181837344). Refactoring / finalize profile format. We'll be sharing profile files more after batch mode lands, so this is the last-call for backward-incompatible changes. (Enso Bot - Apr 14, 2022)
<hr />
**Keziah Wesley** reports a new **STANDUP** for today (2022-04-13):

**Progress:** Refactoring profile format code. It should be finished by 2022-04-15.

**Next Day:** Next day I will be working on the [##181837344 task](https://www.pivotaltracker.com/story/show/181837344). Last changes to profile format and some other small cleanups to get multiprocess-profiling PR ready. (Enso Bot - Apr 14, 2022)
<hr />
**Keziah Wesley** reports a new **STANDUP** for yesterday (2022-04-14):

**Progress:** Refactoring profile format. It should be finished by 2022-04-15.

**Next Day:** Next day I will be working on the [##181837344 task](https://www.pivotaltracker.com/story/show/181837344). Continue on same task. (Enso Bot - Apr 15, 2022)
<hr />
**Keziah Wesley** reports a new **STANDUP** for today (2022-04-15):

**Progress:** The multiprocess-profiling PR was getting too big, so I have split off the core multiprocess changes to be reviewed separately from the deeper profile format changes. It should be finished by 2022-04-15.

**Next Day:** Next day I will be working on the [##181911847 task](https://www.pivotaltracker.com/story/show/181911847). Finishing up profile format changes before batch mode. (Enso Bot - Apr 15, 2022)
<hr />
**Keziah Wesley** reports a new **🔴 DELAY** for yesterday (2022-04-19):

**Summary:** There is **5 days** delay in implementation of the [Profiling: Integrate event logs from application processes besides the IDE (#181837344) task](https://www.pivotaltracker.com/story/show/181837344).
It **will cause 0 days delay** for the delivery of this weekly plan.

**Delay Cause:** Scope expanded a bit. I split off some of the work into #181911847. (Enso Bot - Apr 21, 2022)
<hr />
**Keziah Wesley** reports a new **STANDUP** for yesterday (2022-04-19):

**Progress:** Responding to review. Also implemented basic IDE-to-LS message logging, due to merge conflicts will wait to PR until after Backend Communication Profiling and Multi-process Profiles land. It should be finished by 2022-04-20.

**Next Day:** Next day I will be working on the [##181837344 task](https://www.pivotaltracker.com/story/show/181837344). Finishing multi-process profiles PR. (Enso Bot - Apr 21, 2022)
<hr />
**Keziah Wesley** reports a new **STANDUP** for today (2022-04-20):

**Progress:** Reviewing and responding to review. PR ready for merge. It should be finished by 2022-04-20.

**Next Day:** Next day I will be working on the [##181911847 task](https://www.pivotaltracker.com/story/show/181911847). Finishing up profile format changes before batch mode. (Enso Bot - Apr 21, 2022)
<hr />
	Commit by Kaz Wesley
https://github.com/enso-org/enso/commit/d59710c3cd9e356ce470edfa186c56cdca80f9d9
enso

Multi-process profiles. (#3395)

See: [#181837344](https://www.pivotaltracker.com/story/show/181837344).

I've separated this PR from some deeper changes I'm making to the profile format, because the changeset was getting too complex. The new APIs and tools in this PR are fully-implemented, except the profile format is too simplistic--it doesn't currently support headers that are needed to determine the relative timings of events from different processes.

- Adds basic support for profile files containing data collected by multiple processes.
- Implements `api_events_to_profile`, a tool for converting backend message logs (#3392) to the `profiler` format so they can be merged with frontend profiles (currently they can be merged with `cat`, but the next PR will introduce a merge tool).
- Introduces `message_beanpoles`, a simple tool that diagrams timing relationships between frontend and backend messages.

### Important Notes
- All TODOs introduced here will be addressed in the next PR that defines the new format.
- Introduced a new crate, `enso_profiler_enso_data`, to be used by profile consumers that need to refer to Enso application datatypes to interpret metadata.
- Introduced a `ProfileBuilder` abstraction for writing the JSON profile format; partially decouples the runtime event log structures from the format definition.
- Introducing the conversion performed for `ProfilerBuilder` uncovered that the `.._with_same_start!` low-level `profiler` APIs don't currently work; they return `Started<_>` profilers, but that is inconsistent with the stricter data model that I introduced when I implemented `profiler_data`; they need to return profilers in a created, unstarted state. Low-level async profilers have not been a priority, but once #3382 merges we'll have a way to render their data, which will be really useful because async profilers capture *why* we're doing things. I'll bring up scheduling this in the next performance meeting. (Wojciech Daniło - Apr 21, 2022)
<hr />
